tf_quant_finance/black_scholes/asian_prices.py

```diff
@@ -0,0 +1,218 @@
+      supplied.
+    ValueError: If both `discount_rates` and `discount_factors` is supplied.
+    ValueError: If `is_normal_volatility` is true.
+    ValueError: If last sampling date is later than option expiry date.
+  """
+  if (spots is None) == (forwards is None):
+    raise ValueError('Either spots or forwards must be supplied but not both.')
+  if (discount_rates is not None) and (discount_factors is not None):
+    raise ValueError('At most one of discount_rates and discount_factors may '
+                     'be supplied')
+  if is_normal_volatility:
+    raise ValueError('Cannot price geometric averaging asians analytically '
+                     'under normal volatility')
+  if not np.all(np.maximum(sampling_times[-1], expiries) == expiries):
+    raise ValueError('Sampling times cannot occur after expiry times')
+
+  with tf.name_scope(name or 'discrete_geometric_average_price'):
+    strikes = tf.convert_to_tensor(strikes, dtype=dtype, name='strikes')
+    dtype = strikes.dtype
+    volatilities = tf.convert_to_tensor(
+        volatilities, dtype=dtype, name='volatilities')
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name='expiries')
+
+    if discount_rates is not None:
+      discount_rates = tf.convert_to_tensor(
+          discount_rates, dtype=dtype, name='discount_rates')
+      discount_factors = tf.exp(-discount_rates * expiries)
+    elif discount_factors is not None:
+      discount_factors = tf.convert_to_tensor(
+          discount_factors, dtype=dtype, name='discount_factors')
+      discount_rates = -tf.math.log(discount_factors) / expiries
+    else:
+      discount_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='discount_rates')
+      discount_factors = tf.convert_to_tensor(
+          1.0, dtype=dtype, name='discount_factors')
+
+    if dividend_rates is None:
+      dividend_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='dividend_rates')
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+      spots = forwards * tf.exp(-(discount_rates - dividend_rates) * expiries)
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp((discount_rates - dividend_rates) * expiries)
+
+    if past_fixings is None:
+      running_accumulator = tf.convert_to_tensor(1.0, dtype=dtype)
+      fixing_count = 0
+    else:
+      running_accumulator = tf.reduce_prod(past_fixings, 0)
+      fixing_count = past_fixings.shape[0]
+
+    sample_count = sampling_times.shape[0] + fixing_count
+    sampling_time_forwards = (spots * tf.exp((discount_rates - dividend_rates)
+                                              * sampling_times))
+
+    t1 = tf.reduce_sum(sampling_times, 0) / sample_count
+    t2 = tf.reduce_sum(tf.map_fn(lambda x: tf.minimum(*tf.meshgrid(x, tf.transpose(x))),
+                                 tf.transpose(sampling_times)), [1,2]) / sample_count**2
```

Comment:
here use tf.vectorized_map and set `fallback_to_while_loop` to False. This should automatically vectorize what you want to do and provide significant speed up on a GPU device

Reply:
also, could you please add a comment explaining what is achieved here?

Reply:
Moved to vectorized_map

I've added an explanation comment above - let me know if this works or if you have a standard way of formatting math comments.